replication.go

```diff
@@ -0,0 +1,288 @@
+	if err != nil {
+		return
+	}
+
+	if nparsed != 2 {
+		err = errors.New(fmt.Sprintf("Failed to parsed LSN: %s", lsn))
+		return
+	}
+
+	outputLsn = (upperHalf << 32) + lowerHalf
+	return
+}
+
+// The WAL message contains WAL payload entry data
+type WalMessage struct {
+	// The WAL start position of this data. This
+	// is the WAL position we need to track.
+	WalStart uint64
+	// The server wal end and server time are
+	// documented to track the end position and current
+	// time of the server, both of which appear to be
+	// unimplemented in pg 9.5.
+	ServerWalEnd uint64
+	ServerTime   uint64
+	// The WAL data is the raw unparsed binary WAL entry.
+	// The contents of this are determined by the output
+	// logical encoding plugin.
+	WalData []byte
+}
+
+func (w *WalMessage) Time() time.Time {
+	return time.Unix(0, (int64(w.ServerTime)*1000)+epochNano)
+}
+
+func (w *WalMessage) ByteLag() uint64 {
+	return (w.ServerWalEnd - w.WalStart)
+}
+
+func (w *WalMessage) String() string {
+	return fmt.Sprintf("Wal: %s Time: %s Lag: %d", FormatLSN(w.WalStart), w.Time(), w.ByteLag())
+}
+
+// The server heartbeat is sent periodically from the server,
+// including server status, and a reply request field
+type ServerHeartbeat struct {
+	// The current max wal position on the server,
+	// used for lag tracking
+	ServerWalEnd uint64
+	// The server time, in microseconds since jan 1 2000
+	ServerTime uint64
+	// If 1, the server is requesting a standby status message
+	// to be sent immediately.
+	ReplyRequested byte
+}
+
+func (s *ServerHeartbeat) Time() time.Time {
+	return time.Unix(0, (int64(s.ServerTime)*1000)+epochNano)
+}
+
+func (s *ServerHeartbeat) String() string {
+	return fmt.Sprintf("WalEnd: %s ReplyRequested: %d T: %s", FormatLSN(s.ServerWalEnd), s.ReplyRequested, s.Time())
+}
+
+// The replication message wraps all possible messages from the
+// server received during replication. At most one of the wal message
+// or server heartbeat will be non-nil
+type ReplicationMessage struct {
+	WalMessage      *WalMessage
+	ServerHeartbeat *ServerHeartbeat
+}
+
+// The standby status is the client side heartbeat sent to the postgresql
+// server to track the client wal positions. For practical purposes,
+// all wal positions are typically set to the same value.
+type StandbyStatus struct {
+	// The WAL position that's been locally written
+	WalWritePosition uint64
+	// The WAL position that's been locally flushed
+	WalFlushPosition uint64
+	// The WAL position that's been locally applied
+	WalApplyPosition uint64
+	// The client time in microseconds since jan 1 2000
+	ClientTime uint64
+	// If 1, requests the server to immediately send a
+	// server heartbeat
+	ReplyRequested byte
+}
+
+// Create a standby status struct, which sets all the WAL positions
+// to the given wal position, and the client time to the current time.
+func NewStandbyStatus(walPosition uint64) (status *StandbyStatus) {
```

Comment:
Since we're adding the generic replication concept, we may want to have this NewStandByStatus() accept the Flush/Apply/Write WAL positions as separate values.